WalletWasabi/Models/Label.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace WalletWasabi.Models
+{
+	public class Label : IEquatable<Label>
+	{
+		public static Label Empty { get; } = new Label();
+		public static char[] Separators { get; } = new[] { ',', ':' };
+		public IEnumerable<string> Labels { get; }
+		public bool IsEmpty { get; }
+
+		private string LabelString { get; }
+
+		public Label(params string[] labels) : this(labels as IEnumerable<string>)
+		{
+		}
+
+		public Label(IEnumerable<string> labels)
+		{
+			Labels = (labels
+				   ?.SelectMany(x => x?.Split(Separators, StringSplitOptions.RemoveEmptyEntries) ?? new string[0])
+				   ?.Select(x => x.Trim())
+				   ?.Where(x => x != "")
+				   ?.Distinct(StringComparer.OrdinalIgnoreCase)
+				   ?.OrderBy(x => x)
+				   ?? Enumerable.Empty<string>())
+				   .ToArray();
+
+			HashCode = ((IStructuralEquatable)Labels).GetHashCode(EqualityComparer<string>.Default);
+			IsEmpty = !Labels.Any();
+
+			LabelString = string.Join(", ", Labels);
+		}
+
+		public override string ToString() => LabelString;
+
+		public static Label Merge(IEnumerable<Label> labels)
+		{
+			return new Label(labels?
```

Comment:
Hard to read. I would exclude the edge cases at the beginning. After that I would use `Where(x => !string.IsNullOrEmpty(x))` in the Linq.

Reply:
Not sure how to do this. This code must be as performant as possible (the cluster calculation is a computational bottleneck) so your edge case elimination is not a good approach here.  
I improved the readability however, I recognized we can add null to the Label's constructor. Take a look at if it's good for you: https://github.com/zkSNACKs/WalletWasabi/pull/2201/commits/67d9c3ba83d416aebb196ac59733c553b792b2d7

Reply:
That is better. Go with it.